ambersim/base.py

```diff
@@ -0,0 +1,100 @@
+import jax
+from jax import numpy as jp
+import numpy as np
+from typing import Tuple, Dict, Any
+
+from brax.base import Base, Motion, Transform
+from brax.envs.base import Env
+import mujoco
+from mujoco import mjx
+from flax import struct
+
+
+class MjxEnv(Env):
+    """API for driving an MJX system for training and inference in brax."""
+
+    def __init__(
+        self,
+        mj_model: mujoco.MjModel,
+        physics_steps_per_control_step: int = 1,
+    ):
+        """Initializes MjxEnv.
+
+        Args:
+          mj_model: mujoco.MjModel
+          physics_steps_per_control_step: the number of times to step the physics
+            pipeline for each environment step
+        """
+        self.model = mj_model
+        self.data = mujoco.MjData(mj_model)
+        self.sys = mjx.device_put(mj_model)
+        self._physics_steps_per_control_step = physics_steps_per_control_step
+
+    def pipeline_init(self, qpos: jax.Array, qvel: jax.Array) -> mjx.Data:
+        """Initializes the physics state."""
+        data = mjx.device_put(self.data)
+        data = data.replace(qpos=qpos, qvel=qvel, ctrl=jp.zeros(self.sys.nu))
+        data = mjx.forward(self.sys, data)
+        return data
+
+    def pipeline_step(self, data: mjx.Data, ctrl: jax.Array) -> mjx.Data:
+        """Takes a physics step using the physics pipeline."""
+
+        def f(data, _):
+            data = data.replace(ctrl=ctrl)
+            return (
+                mjx.step(self.sys, data),
+                None,
+            )
+
+        data, _ = jax.lax.scan(f, data, (), self._physics_steps_per_control_step)
+        return data
+
+    @property
+    def dt(self) -> jax.Array:
+        """The timestep used for each env step."""
+        return self.sys.opt.timestep * self._physics_steps_per_control_step
+
+    @property
+    def observation_size(self) -> int:
```

Comment:
nit: docstring

Reply:
Oh dang I missed that this was public -- personally I think this needs a docstring (especially in a base class) so auto-generated docs will look good

Reply:
good point on autogen docs. i'll open an issue for implementing them - i've actually never done it.